causal_testing/testing/estimators.py

```diff
@@ -100,16 +101,97 @@ def __init__(
         df: pd.DataFrame = None,
         effect_modifiers: dict[str:Any] = None,
         formula: str = None,
+        alpha: float = 0.05,
     ):
-        super().__init__(treatment, treatment_value, control_value, adjustment_set, outcome, df, effect_modifiers)
+        super().__init__(
+            treatment, treatment_value, control_value, adjustment_set, outcome, df, effect_modifiers, alpha=alpha
+        )
 
-        self.model = None
+        if effect_modifiers is None:
+            effect_modifiers = []
 
         if formula is not None:
             self.formula = formula
+
+        else:
+            terms = [treatment] + sorted(list(adjustment_set)) + sorted(list(effect_modifiers))
+            self.formula = f"{outcome} ~ {'+'.join(terms)}"
+
+    @abstractmethod
+    def add_modelling_assumptions(self):
+        """
+        Add modelling assumptions to the estimator. This is a list of strings which list the modelling assumptions that
+        must hold if the resulting causal inference is to be considered valid.
+        """
+
+    def get_terms_from_formula(self) -> tuple[str, str, list[str]]:
+        """
+        Parse all the terms from a Patsy formula string into outcome, treatment and covariate variables.
+
+        Formulae are expected to only have a single left hand side term.
+
+        :return: a truple containing the outcome, treatment and covariate variable names in string format
+        """
+        desc = ModelDesc.from_formula(self.formula)
+        if len(desc.lhs_termlist) > 1:
+            raise ValueError("More than 1 left hand side term provided in formula, only single term is accepted")
+        outcome = desc.lhs_termlist[0].factors[0].code
+        rhs_terms = set()
+        for term in desc.rhs_termlist:
+            if term.factors:
+                rhs_terms.add(term.factors[0].code)
+        if self.treatment not in rhs_terms:
+            raise ValueError(f"Treatment variable '{self.treatment}' not found in formula")
+        rhs_terms.remove(self.treatment)
+        covariates = rhs_terms
+        if covariates is None:
```

Comment:
Will this ever actually come about? Surely there should always be at least one RHS term, even if it's just the treatment?

Reply:
The code link isn't too clear on if you are talking about the ValueError or the `if covariates is None`,

The `if covariates is None:` addresses the case where there is only the treatment variable on the RHS and therefore there are no covariates, but the `constructive_backdoor_criterion` expects a list instead of `None`. 

The `ValueError` covers the case where the formula does not include the treatment hopefully.

Let me know if that clears it up?

Reply:
Ah sorry. I was going for `if covariates is None`, and that makes sense. Again, a test to cover this would probably be good. Causal DAG `X -> Y` with formula `Y ~ X` should do it